File: sync-problems/cpp/h2o/main.cpp
// H2O Water Factory — starter template (C++20)
//
// Scenario (lecture T7.1): hydrogen and oxygen atoms arrive at a factory.
// When 2 H + 1 O are present, they bond into a water molecule.
//
// Build (from sync-problems/cpp/):
//   cmake -S . -B build && cmake --build build
//   ./build/h2o/h2o
//
// Strategies worth trying (lecture):
//   1. WaterFactory1 (BUGGY): just std::barrier<>{3}. Catches the count
//      but NOT the type — 3 oxygens can bond into ozone.
//   2. WaterFactory2 (BUGGY): std::counting_semaphore<>{2}/{1} for H/O.
//      Catches the type but lets atoms bond solo without 3 present.
//   3. WaterFactory3 (CORRECT): combine semaphores AND barrier — semaphores
//      gate the type counts; barrier gates the start of bonding.
//
// Invariants checked by the harness:
//   - h_in_bond <= 2 always (catches ozone)
//   - o_in_bond <= 1 always (catches double-oxygen)
//   - max_total_in_bond should reach 3 (catches "no barrier" — atoms bonding solo)
//   - h_total_bonds == 2 * N_MOLECULES, o_total_bonds == N_MOLECULES

#include <atomic>
#include <cassert>
#include <chrono>
#include <iostream>
#include <random>
#include <thread>
#include <vector>
#include <semaphore>
#include <barrier>
using namespace std::chrono_literals;

constexpr int N_MOLECULES = 50;
constexpr int N_HYDROGEN = 2 * N_MOLECULES;
constexpr int N_OXYGEN = N_MOLECULES;

// ----- Invariant tracking (do not touch) -----
std::atomic<int> h_in_bond{0};
std::atomic<int> o_in_bond{0};
std::atomic<int> max_total_in_bond{0};
std::atomic<int> h_total_bonds{0};
std::atomic<int> o_total_bonds{0};

void update_max_total() {
    int total = h_in_bond.load() + o_in_bond.load();
    int cur = max_total_in_bond.load();
    while (total > cur && !max_total_in_bond.compare_exchange_weak(cur, total)) {}
}

void bond_h(int id) {
    int h = h_in_bond.fetch_add(1) + 1;
    assert(h <= 2 && "ozone! more than 2 H atoms in bond simultaneously");
    update_max_total();
    std::this_thread::sleep_for(2ms);
    update_max_total();
    std::this_thread::sleep_for(1ms);
    h_in_bond.fetch_sub(1);
    h_total_bonds.fetch_add(1);
    (void)id;
}

void bond_o(int id) {
    int o = o_in_bond.fetch_add(1) + 1;
    assert(o <= 1 && "more than 1 O in bond simultaneously");
    update_max_total();
    std::this_thread::sleep_for(2ms);
    update_max_total();
    std::this_thread::sleep_for(1ms);
    o_in_bond.fetch_sub(1);
    o_total_bonds.fetch_add(1);
    (void)id;
}

// ==========================================================================
// TODO: Implement WaterFactory.
//
// API contract:
//   - hydrogen(id): block until grouped with 1 H + 1 O, then call bond_h(id).
//   - oxygen(id):   block until grouped with 2 H,         then call bond_o(id).
//
// Strategies (pick one, then try a second for comparison):
//   1. std::counting_semaphore<>{2} (H) + std::counting_semaphore<>{1} (O)
//      + std::barrier<>{3}. Each acquirer holds its semaphore permit ACROSS
//      the barrier+bond, then releases on exit. Mirrors WaterFactory3.
//   2. std::mutex + std::condition_variable + state machine. Wake the right
//      type when the right counts (h_waiting >= 2 && o_waiting >= 1) are met.
//   3. (Compare to) std::latch<3>: like barrier but single-use. What goes wrong
//      if you use a latch here? (hint: only one molecule can ever bond.)
// ==========================================================================
class WaterFactory {
public:
    void hydrogen(int id) {
        // TODO: coordinate, then bond_h(id)
        hydrogen_sem_.acquire();
        barrier.arrive_and_wait();
        bond_h(id);
        hydrogen_sem_.release();
    }

    void oxygen(int id) {
        // TODO: coordinate, then bond_o(id)
        oxygen_sem_.acquire();
        barrier.arrive_and_wait();
        bond_o(id);
        oxygen_sem_.release();
    }

private:
    // TODO: fields
    std::counting_semaphore<> oxygen_sem_{1};
    std::counting_semaphore<> hydrogen_sem_{2};
    std::barrier<> barrier{3};
};

int main() {
    WaterFactory factory;
    std::vector<std::thread> threads;
    threads.reserve(N_HYDROGEN + N_OXYGEN);

    std::mt19937 rng(42);
    std::uniform_int_distribution<int> jitter(0, 5);

    auto start = std::chrono::steady_clock::now();
    for (int i = 0; i < N_HYDROGEN; ++i) {
        std::this_thread::sleep_for(std::chrono::milliseconds(jitter(rng)));
        threads.emplace_back([&factory, i] { factory.hydrogen(i); });
    }
    for (int i = 0; i < N_OXYGEN; ++i) {
        std::this_thread::sleep_for(std::chrono::milliseconds(jitter(rng)));
        threads.emplace_back([&factory, i] { factory.oxygen(i); });
    }
    for (auto& t : threads) t.join();
    auto elapsed = std::chrono::steady_clock::now() - start;

    std::cout << "H bonds: " << h_total_bonds.load() << "/" << N_HYDROGEN
              << "  O bonds: " << o_total_bonds.load() << "/" << N_OXYGEN
              << "  max simultaneous in bond: " << max_total_in_bond.load()
              << "  time: "
              << std::chrono::duration_cast<std::chrono::milliseconds>(elapsed).count()
              << "ms\n";

    bool ok = true;
    if (h_total_bonds.load() != N_HYDROGEN) {
        std::cerr << "MISSING H BONDS!\n"; ok = false;
    }
    if (o_total_bonds.load() != N_OXYGEN) {
        std::cerr << "MISSING O BONDS!\n"; ok = false;
    }
    if (max_total_in_bond.load() != 3) {
        std::cerr << "max_total_in_bond should be 3, got "
                  << max_total_in_bond.load()
                  << " — atoms are bonding without all 3 being present\n";
        ok = false;
    }
    return ok ? 0 : 1;
}

File: sync-problems/rust/h2o/src/main.rs
// H2O Water Factory — tokio template (Rust)
//
// Scenario (lecture T7.1): hydrogen and oxygen atoms arrive at a factory.
// When 2 H + 1 O are present, they bond into a water molecule.
//
// Run:
//   cargo run -p h2o --release
//
// Strategies (pick one, then try a second):
//   1. tokio::sync::Semaphore::new(2) (H) + Semaphore::new(1) (O)
//      + tokio::sync::Barrier::new(3). Mirrors C++ WaterFactory3.
//      Each acquirer holds its semaphore permit ACROSS the barrier+bond,
//      then releases on exit. Use `permit.forget()` + `add_permits(1)` if
//      you want classical-Dijkstra semantics (see barbershop notes).
//   2. Daemon task with mpsc channels (Go-style "two-phase commit"):
//        precommit: 2 H + 1 O each send their oneshot::Sender to the daemon
//        commit:    daemon sends () to all 3
//        postcommit: each atom sends () back; daemon waits for all 3
//      The daemon owns the protocol; atoms just submit themselves and wait.
//   3. Leader election with oxygen as leader (Go-style with a Mutex/permit).
//      Oxygen takes a 1-permit "leader" semaphore; collects 2 H precommits
//      via a shared mpsc; sends them go; bonds; waits for both H to finish;
//      releases the leader permit.
//
// Invariants enforced by the harness:
//   - H_IN_BOND <= 2  (assertion fires on increment past 2 — catches ozone)
//   - O_IN_BOND <= 1  (catches double-oxygen)
//   - MAX_TOTAL_IN_BOND should reach 3 (catches "no barrier" — atoms bonding solo)
//   - H_TOTAL_BONDS == 2 * N_MOLECULES, O_TOTAL_BONDS == N_MOLECULES

use rand::Rng;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::{Semaphore, TryAcquireError, Barrier};


const N_MOLECULES: usize = 50;
const N_HYDROGEN: usize = 2 * N_MOLECULES;
const N_OXYGEN: usize = N_MOLECULES;

// ----- Invariant tracking (do not touch) -----
static H_IN_BOND: AtomicUsize = AtomicUsize::new(0);
static O_IN_BOND: AtomicUsize = AtomicUsize::new(0);
static MAX_TOTAL_IN_BOND: AtomicUsize = AtomicUsize::new(0);
static H_TOTAL_BONDS: AtomicUsize = AtomicUsize::new(0);
static O_TOTAL_BONDS: AtomicUsize = AtomicUsize::new(0);

fn update_max_total() {
    let total = H_IN_BOND.load(Ordering::SeqCst) + O_IN_BOND.load(Ordering::SeqCst);
    let mut cur = MAX_TOTAL_IN_BOND.load(Ordering::SeqCst);
    while total > cur {
        match MAX_TOTAL_IN_BOND.compare_exchange(cur, total, Ordering::SeqCst, Ordering::SeqCst) {
            Ok(_) => break,
            Err(actual) => cur = actual,
        }
    }
}

async fn bond_h(_id: usize) {
    let h = H_IN_BOND.fetch_add(1, Ordering::SeqCst) + 1;
    assert!(h <= 2, "ozone! {} H atoms in bond simultaneously", h);
    update_max_total();
    tokio::time::sleep(Duration::from_millis(2)).await;
    update_max_total();
    tokio::time::sleep(Duration::from_millis(1)).await;
    H_IN_BOND.fetch_sub(1, Ordering::SeqCst);
    H_TOTAL_BONDS.fetch_add(1, Ordering::SeqCst);
}

async fn bond_o(_id: usize) {
    let o = O_IN_BOND.fetch_add(1, Ordering::SeqCst) + 1;
    assert!(o <= 1, "more than 1 O in bond ({} O atoms)", o);
    update_max_total();
    tokio::time::sleep(Duration::from_millis(2)).await;
    update_max_total();
    tokio::time::sleep(Duration::from_millis(1)).await;
    O_IN_BOND.fetch_sub(1, Ordering::SeqCst);
    O_TOTAL_BONDS.fetch_add(1, Ordering::SeqCst);
}

// ==========================================================================
// TODO: Implement WaterFactory.
//
// API contract:
//   hydrogen(id).await: block until grouped with 1 H + 1 O, then call bond_h(id).await
//   oxygen(id).await:   block until grouped with 2 H,         then call bond_o(id).await
// ==========================================================================
pub struct WaterFactory {
    // TODO: fields
    oxygen_sem: Semaphore,
    hydrogen_sem: Semaphore,
    barrier: Barrier,
    
}

impl WaterFactory {
    pub fn new() -> Self {
        Self {
            oxygen_sem: Semaphore::new(1),
            hydrogen_sem: Semaphore::new(2),
            barrier: Barrier::new(3),
        }
    }

    pub async fn hydrogen(&self, id: usize) {
        // TODO: coordinate, then bond_h(id).await
        self.hydrogen_sem.acquire().await.unwrap().forget();
        self.barrier.wait().await;
        bond_h(id).await;
        self.hydrogen_sem.add_permits(1);
    }

    pub async fn oxygen(&self, id: usize) {
        // TODO: coordinate, then bond_o(id).await\
        self.oxygen_sem.acquire().await.unwrap().forget();
        self.barrier.wait().await;
        bond_o(id).await;
        self.oxygen_sem.add_permits(1);
    }
}

#[tokio::main]
async fn main() {
    let factory = Arc::new(WaterFactory::new());
    let mut handles = Vec::new();
    let mut rng = rand::thread_rng();

    let start = std::time::Instant::now();
    for i in 0..N_HYDROGEN {
        tokio::time::sleep(Duration::from_millis(rng.gen_range(0..5))).await;
        let f = Arc::clone(&factory);
        handles.push(tokio::spawn(async move { f.hydrogen(i).await }));
    }
    for i in 0..N_OXYGEN {
        tokio::time::sleep(Duration::from_millis(rng.gen_range(0..5))).await;
        let f = Arc::clone(&factory);
        handles.push(tokio::spawn(async move { f.oxygen(i).await }));
    }
    for h in handles {
        h.await.unwrap();
    }
    let elapsed = start.elapsed();

    let h_bonds = H_TOTAL_BONDS.load(Ordering::SeqCst);
    let o_bonds = O_TOTAL_BONDS.load(Ordering::SeqCst);
    let max_total = MAX_TOTAL_IN_BOND.load(Ordering::SeqCst);

    println!(
        "H bonds: {}/{}  O bonds: {}/{}  max simultaneous in bond: {}  time: {:?}",
        h_bonds, N_HYDROGEN, o_bonds, N_OXYGEN, max_total, elapsed
    );

    let mut ok = true;
    if h_bonds != N_HYDROGEN {
        eprintln!("MISSING H BONDS");
        ok = false;
    }
    if o_bonds != N_OXYGEN {
        eprintln!("MISSING O BONDS");
        ok = false;
    }
    if max_total != 3 {
        eprintln!("max_total_in_bond should be 3 — atoms are bonding without all 3 being present");
        ok = false;
    }
    if !ok {
        std::process::exit(1);
    }
}
